node-manager/src/socket.rs

```diff
@@ -0,0 +1,413 @@
+            WsSocketDescriptor::Mutiny(s) => s.disconnect_socket(),
+        }
+    }
+}
+
+pub(crate) struct WsTcpSocketDescriptor {
+    conn: Arc<Proxy>,
+    id: u64,
+}
+impl WsTcpSocketDescriptor {
+    pub fn new(conn: Arc<Proxy>) -> Self {
+        let id = ID_COUNTER.fetch_add(1, Ordering::AcqRel);
+        Self { conn, id }
+    }
+}
+
+impl ReadDescriptor for WsTcpSocketDescriptor {
+    async fn read(&self) -> Option<Result<Message, gloo_net::websocket::WebSocketError>> {
+        self.conn.read.lock().await.next().await
+    }
+}
+
+unsafe impl Send for WsTcpSocketDescriptor {}
+unsafe impl Sync for WsTcpSocketDescriptor {}
+
+impl peer_handler::SocketDescriptor for WsTcpSocketDescriptor {
+    fn send_data(&mut self, data: &[u8], _resume_read: bool) -> usize {
+        let vec = Vec::from(data);
+        self.conn.send(vec);
+        data.len()
+    }
+
+    fn disconnect_socket(&mut self) {
+        let cloned = self.conn.write.clone();
+        spawn_local(async move {
+            let mut conn = cloned.lock().await;
+            let _ = conn.close();
+            debug!("closed websocket");
+        });
+    }
+}
+impl Clone for WsTcpSocketDescriptor {
+    fn clone(&self) -> Self {
+        Self {
+            conn: Arc::clone(&self.conn),
+            id: self.id,
+        }
+    }
+}
+impl Eq for WsTcpSocketDescriptor {}
+impl PartialEq for WsTcpSocketDescriptor {
+    fn eq(&self, o: &Self) -> bool {
+        self.id == o.id
+    }
+}
+impl Hash for WsTcpSocketDescriptor {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.id.hash(state);
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct MultiWsSocketDescriptor {
+    conn: Arc<Proxy>,
+    read_from_sub_socket: Receiver<Message>,
+    send_to_multi_socket: Sender<Message>,
+    socket_map: Arc<Mutex<HashMap<Vec<u8>, Sender<Message>>>>,
+    peer_manager: Arc<PeerManager>,
+    connected: Arc<AtomicBool>,
+}
+impl MultiWsSocketDescriptor {
+    pub fn new(conn: Arc<Proxy>, peer_manager: Arc<PeerManager>) -> Self {
+        debug!("setting up multi websocket descriptor");
+
+        let (send_to_multi_socket, read_from_sub_socket): (Sender<Message>, Receiver<Message>) =
+            unbounded();
+
+        let socket_map: Arc<Mutex<HashMap<Vec<u8>, Sender<Message>>>> =
+            Arc::new(Mutex::new(HashMap::new()));
+
+        Self {
+            conn,
+            send_to_multi_socket,
+            read_from_sub_socket,
+            socket_map,
+            peer_manager,
+            connected: Arc::new(AtomicBool::new(true)),
+        }
+    }
+
+    pub fn connected(&self) -> bool {
+        self.connected.load(Ordering::Relaxed)
+    }
+
+    pub fn reconnect(&mut self, conn: Arc<Proxy>) {
+        let (send_to_multi_socket, read_from_sub_socket): (Sender<Message>, Receiver<Message>) =
+            unbounded();
+
+        let socket_map: Arc<Mutex<HashMap<Vec<u8>, Sender<Message>>>> =
+            Arc::new(Mutex::new(HashMap::new()));
+
```

Comment:
I think this is one problem i need to explore. I shouldn't create a new map, but delete the old maps and sub sockets..